src/reducers/vms.js

```diff
@@ -28,66 +27,58 @@ import {
 import { actionReducer, removeMissingItems } from './utils'
 import { sortFields } from '_/utils'
 
-const EMPTY_MAP = Immutable.fromJS({})
-const EMPTY_ARRAY = Immutable.fromJS([])
-
 const initialState = Immutable.fromJS({
   vms: {},
   pools: {},
-  filters: EMPTY_MAP,
+  filters: {},
   sort: { ...sortFields[0], isAsc: true },
 
-  page: 1,
-
   missedVms: Immutable.Set(),
 
-  /**
-   * true ~ we need to fetch further vms and pools
-   * false ~ all visible entities already fetched
-   */
-  notAllPagesLoaded: true,
+  vmsPage: 0,
+  vmsExpectMorePages: true,
+
+  poolsPage: 0,
+  poolsExpectMorePages: true,
 
   correlationResult: {},
 })
 
 const vms = actionReducer(initialState, {
-  [UPDATE_VMS] (state, { payload: { vms, copySubResources, page } }) {
+
+  // vms come in as Internal transformed JS objects that will be pushed to ImmutableJS
+  //     objects after an optional merge with existing vm data
+  [UPDATE_VMS] (state, { payload: { vms, copySubResources } }) {
     const updates = {}
 
     vms.forEach(vm => {
-      if (!state.getIn(['vms', vm.id])) {
-        state = state.set('notAllPagesLoaded', true)
-      }
+      const existingVm = state.hasIn(['vms', vm.id]) ? state.getIn(['vms', vm.id]).toJS() : false
```

Comment:
Paging logic might get confused if new page contains existing vm's. In worst case no new vm's can be fetched i.e. if vm's were added to pages before newly fetched page or some vm's were renamed(vm's are fetched sorted by name).
Example of such logic:
```
function* currentVmsIds ({ payload: { count, page } }) {
    const start = count * page
    const end = start + count
```

Reply:
Refresh pull ALL pages in one call.  That would cover the edge case of "page size is 20, I fetch my 5 VMs and someone then creates VM number 6".

Only scrolling the infinite scroll does pages fetching.

Reply:
the next line in the example above is:
```
return Array.from(yield select(state => state.vms.get('pools').keys())).slice(start, end)
```
I was concerned about going past the end of array which i.e. Java would result in ArrayIndexOutOfBoundsException.
JS seems to be very relaxed in this aspect and the above will work also for invalid start/end so in this place we are safe.
However in general similar logic should rather be based on actual number of vms.

Reply:
The only thing the subset of VM or Pool ids is being used for, in the `fetchByPage` saga, is to count how many things were fetched.  If enough things were fetched to fill a single page of things, we then expect more pages.

The logic is simply based on counts of things in a fetch page.  Contents of the fetch pages don't matter.  We only assume there are more pages if the current page is full.

And yeah, in JavaScript, this is perfectly fine and will return `[]`:

```jsx
   [ 1, 3, 5, 7, 9 ].slice(100, 200)
```

Aside, I did always get annoyed at fun little features of dynamic languages things like:
```jsx
    const a = []
    a.length // === 0
    a[100] = 'b'
    a.length // === 101
```